src/routes/maven.rs

```diff
@@ -0,0 +1,262 @@
+use crate::auth::get_user_from_headers;
+use crate::database;
+use crate::models::mods::ModId;
+use crate::routes::ApiError;
+use actix_web::{get, web, HttpRequest, HttpResponse};
+use sqlx::PgPool;
+use yaserde_derive::YaSerialize;
+
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(root = "metadata", rename = "metadata")]
+pub struct Metadata {
+    #[yaserde(rename = "groupId")]
+    group_id: String,
+    #[yaserde(rename = "artifactId")]
+    artifact_id: String,
+    versioning: Versioning,
+}
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(rename = "versioning")]
+pub struct Versioning {
+    latest: String,
+    release: String,
+    versions: Versions,
+    #[yaserde(rename = "lastUpdated")]
+    last_updated: String,
+}
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(rename = "versions")]
+pub struct Versions {
+    #[yaserde(rename = "version")]
+    versions: Vec<String>,
+}
+#[derive(Default, Debug, Clone, YaSerialize)]
+#[yaserde(rename = "project", namespace = "http://maven.apache.org/POM/4.0.0")]
+pub struct MavenPom {
+    #[yaserde(rename = "xsi:schemaLocation", attribute)]
+    schema_location: String,
+    #[yaserde(rename = "xmlns:xsi", attribute)]
+    xsi: String,
+    #[yaserde(rename = "modelVersion")]
+    model_version: String,
+    #[yaserde(rename = "groupId")]
+    group_id: String,
+    #[yaserde(rename = "artifactId")]
+    artifact_id: String,
+    version: String,
+    name: String,
+    description: String,
+}
+
+#[get("maven/modrinth/{id}/maven-metadata.xml")]
+pub async fn maven_metadata(
+    req: HttpRequest,
+    info: web::Path<(String,)>,
+    pool: web::Data<PgPool>,
+) -> Result<HttpResponse, ApiError> {
+    let string = info.into_inner().0;
+    let id_option: Option<ModId> = serde_json::from_str(&*format!("\"{}\"", string)).ok();
+
+    let mut mod_data;
+
+    if let Some(id) = id_option {
+        mod_data = database::models::Mod::get_full(id.into(), &**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+        if mod_data.is_none() {
+            mod_data = database::models::Mod::get_full_from_slug(&string, &**pool)
+                .await
+                .map_err(|e| ApiError::DatabaseError(e.into()))?;
+        }
+    } else {
+        mod_data = database::models::Mod::get_full_from_slug(&string, &**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
+    }
+
+    let user_option = get_user_from_headers(req.headers(), &**pool).await.ok();
+
+    if let Some(data) = mod_data {
+        let mut authorized = !data.status.is_hidden();
+
+        if let Some(user) = user_option {
```

Comment:
That kind of functions should have its own function, as it's really possible ti will be needed somewhere else in the code, and will also allow us to reduce the identation in this part of the code.